lib/tool_shed/util/repository_util.py

```diff
@@ -198,41 +194,37 @@ def create_repository(
         long_description=long_description,
         user_id=user_id,
     )
-    # Flush to get the id.
     sa_session.add(repository)
-    session = sa_session()
-    with transaction(session):
-        session.commit()
-    # Create an admin role for the repository.
-    create_repository_admin_role(app, repository)
-    # Determine the repository's repo_path on disk.
-    dir = os.path.join(app.config.file_path, *util.directory_hash_id(repository.id))
-    # Create directory if it does not exist.
-    if not os.path.exists(dir):
-        os.makedirs(dir)
-    # Define repo name inside hashed directory.
-    repository_path = os.path.join(dir, "repo_%d" % repository.id)
-    # Create local repository directory.
-    if not os.path.exists(repository_path):
-        os.makedirs(repository_path)
-    # Create the local repository.
-    init_repository(repo_path=repository_path)
-    # Add an entry in the hgweb.config file for the local repository.
-    lhs = f"{app.config.hgweb_repo_prefix}{repository.user.username}/{repository.name}"
-    app.hgweb_config_manager.add_entry(lhs, repository_path)
-    # Create a .hg/hgrc file for the local repository.
-    create_hgrc_file(app, repository)
-    flush_needed = False
     if category_ids:
         # Create category associations
         for category_id in category_ids:
             category = sa_session.query(app.model.Category).get(app.security.decode_id(category_id))
             rca = app.model.RepositoryCategoryAssociation(repository, category)
             sa_session.add(rca)
-            flush_needed = True
-    if flush_needed:
-        with transaction(session):
-            session.commit()
+    # Create an admin role for the repository.
+    create_repository_admin_role(app, repository)
+    # Create a temporary repo_path on disk.
+    with tempfile.TemporaryDirectory(
+        dir=app.config.file_path, prefix="f{repository.user.username}-{repository.name}"
+    ) as repository_path:
+        # Create the local repository.
+        init_repository(repo_path=repository_path)
+        # Create a .hg/hgrc file for the local repository.
+        create_hgrc_file(app, repository)
+        # Add an entry in the hgweb.config file for the local repository.
+        lhs = f"{app.config.hgweb_repo_prefix}{repository.user.username}/{repository.name}"
+        # Flush to get the id.
+    session = sa_session()
+    with transaction(session):
+        session.commit()
```

Comment:
The proper fix would be to use a uuid and only commit after everything else has worked properly, but I guess this already reduces the surface area for problems considerably, and we can migrate to UUID against dev.